engine/src/main/java/org/hibernate/validator/messageinterpolation/AbstractMessageInterpolator.java

```diff
@@ -251,10 +251,11 @@ public String interpolate(String message, Context context, Locale locale) {
 	 */
 	private String interpolateMessage(String message, Context context, Locale locale)
 			throws MessageDescriptorFormatException {
-		LocalizedMessage localisedMessage = new LocalizedMessage( message, locale );
+		LocalizedMessage localisedMessage = null;
 		String resolvedMessage = null;
 
 		if ( cachingEnabled ) {
+			localisedMessage = new LocalizedMessage( message, locale );
```

Comment:
Mmmh, this is a bold move... I don't like this pattern at all as the compiler is all lost.

I think the issue comes from the fact that we didn't have Java 8 at this time.

What about:
```java
		String resolvedMessage = null;

		if ( cachingEnabled ) {
			resolvedMessage = resolvedMessages.computeIfAbsent( new LocalizedMessage( message, locale ), lm -> resolveMessage( message, locale ) );
		} else {
			resolvedMessage = resolveMessage( message, locale );
		}
```

with `resolveMessage()` being:

```java
private String resolveMessage(String message, Locale locale) {
		String resolvedMessage = message;

		ResourceBundle userResourceBundle = userResourceBundleLocator
				.getResourceBundle( locale );

		ResourceBundle constraintContributorResourceBundle = contributorResourceBundleLocator
				.getResourceBundle( locale );

		ResourceBundle defaultResourceBundle = defaultResourceBundleLocator
				.getResourceBundle( locale );

		String userBundleResolvedMessage;
		boolean evaluatedDefaultBundleOnce = false;
		do {
			// search the user bundle recursive (step1)
			userBundleResolvedMessage = interpolateBundleMessage(
					resolvedMessage, userResourceBundle, locale, true
			);

			// search the constraint contributor bundle recursive (only if the user did not define a message)
			if ( !hasReplacementTakenPlace( userBundleResolvedMessage, resolvedMessage ) ) {
				userBundleResolvedMessage = interpolateBundleMessage(
						resolvedMessage, constraintContributorResourceBundle, locale, true
				);
			}

			// exit condition - we have at least tried to validate against the default bundle and there was no
			// further replacements
			if ( evaluatedDefaultBundleOnce
					&& !hasReplacementTakenPlace( userBundleResolvedMessage, resolvedMessage ) ) {
				break;
			}

			// search the default bundle non recursive (step2)
			resolvedMessage = interpolateBundleMessage(
					userBundleResolvedMessage,
					defaultResourceBundle,
					locale,
					false
			);
			evaluatedDefaultBundleOnce = true;
		} while ( true );
		
		return resolvedMessage;
	}
```

(we probably need to rework the comments with the steps)

Reply:
+1 that would be better

Reply:
Mmmh, and we could probably also use `computeIfAbsent()` for `tokenizedParameterMessages` and `tokenizedELMessages`. It will make the code more readable.